x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -304,14 +320,14 @@ public void writeTo(StreamOutput out) throws IOException {
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
-        doXContentBody(builder, params);
-        builder.endObject();
-        return builder;
-    }
-
-    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
         builder.field(ID.getPreferredName(), id);
         builder.field(Job.ID.getPreferredName(), jobId);
+        // Config type field is added for the migration to index documents
+        // and isn't needed in cluster state configs. Not writing the field
+        // protects against BWC issues.
+        if (params.paramAsBoolean(ToXContentParams.FOR_CLUSTER_STATE, false) == false) {
```

Comment:
Originally this `ToXContentParams.FOR_CLUSTER_STATE` parameter was used to avoid showing the user internal implementation details.  Based on the original intent the parameter should probably have been named `ToXContentParams.FOR_INTERNAL_STORAGE`.

There is also another constant, `MlMetaIndex.INCLUDE_TYPE_KEY`, which is used to avoid showing the end user the metadata document type in REST responses.  I think we should also use that for config documents.  So maybe move `INCLUDE_TYPE_KEY` into `ToXContentParams` so it can be reused here.

Then the logic here will be to include the type neither for cluster state nor for REST responses.